0002-RCBS-Crypto-building-block.md

```diff
@@ -0,0 +1,675 @@
+message SubtleVerifyResponse {
+  // True if the signature is valid.
+  bool valid = 1;
+}
+
+// rpc Encrypt(stream EncryptRequest) returns (stream EncryptResponse);
+
+message EncryptRequest {
+  // Request details. Must be present in the first message only.
+  EncryptRequestOptions options = 1;
+  // Chunk of data of arbitrary size.
+  common.v1.StreamPayload payload = 2;
+}
+
+message EncryptRequestOptions {
+  // Name of the component
+  string component_name = 1 [json_name="component"];
+  // Name (or name/version) of the key.
+  string key = 2;
+  // Force algorithm to use: "aes-gcm" or "chacha20-poly1305" (optional)
+  string algorithm = 3;
+}
+
+message EncryptResponse {
+  // Chunk of data.
+  common.v1.StreamPayload payload = 1;
+}
+
+// rpc Decrypt(stream DecryptRequest) returns (stream DecryptResponse);
+
+message DecryptRequest {
+  // Request details. Must be present in the first message only.
+  DecryptRequestOptions options = 1;
+  // Chunk of data of arbitrary size.
+  common.v1.StreamPayload payload = 2;
+}
+
+message DecryptRequestOptions {
+  // Name of the component
+  string component_name = 1 [json_name="component"];
+  // Name (or name/version) of the key.
+  string key = 2;
+}
+
+message DecryptResponse {
+  // Chunk of data.
+  common.v1.StreamPayload payload = 1;
+}
+```
+
+> For the `common.v1.StreamPayload` message, see [dapr/dapr#5170](https://github.com/dapr/dapr/pull/5170)
+
+The `Encrypt` and `Decrypt` methods are stream-based. Dapr will read from the client until it has sufficient data, and will then send back the encrypted/decrypted data to the client. Clients must thus both send data to the RPC and listen for incoming messages. SDKs can offer to consumer methods to read the data as a stream (e.g. in Go, they accept an `io.Reader` and return an `io.Reader`)
+
+## HTTP APIs
+
+The HTTP APIs are developed in a way that is the exact "port" of the gRPC "subtle" APIs, and the contents of the request and response bodies match exactly the fields in the gRPC APIs (except for the component name).
+
+Methods are:
+
+- `POST /v1.0/subtlecrypto/[component]/getkey` -> SubtleGetKey
+- `POST /v1.0/subtlecrypto/[component]/encrypt` -> SubtleEncrypt
+- `POST /v1.0/subtlecrypto/[component]/decrypt` -> SubtleDecrypt
+- `POST /v1.0/subtlecrypto/[component]/wrapkey` -> SubtleWrapKey
+- `POST /v1.0/subtlecrypto/[component]/unwrapkey` -> SubtleUnwrapKey
+- `POST /v1.0/subtlecrypto/[component]/sign` -> SubtleSign
+- `POST /v1.0/subtlecrypto/[component]/verify` -> SubtleVerify
+
+> Note: URL will begin with `/v1.0-alpha1` while in preview
+
+**Currently, higher-level APIs are not offered via HTTP**. This is because it's not possible, using HTTP, to have a stream where Dapr sends back data to the client before the client's request is complete. Without that, we'd have to buffer the entire data in-memory in the runtime, which would be not practical when encrypting very large files.
```

Comment:
Same comment here. We should offer that for compatibility reasons even with the perf penalty. Make it clear that for higher performance, should use gRPC with the SDK preferably.